File: frontend/src/types/multiplayer.ts
import { z } from 'zod';

import { chatPayloadSchema, sessionBindingsSchema } from '@/types/domain';

export const multiplayerEventTypeValues = [
  'room_status',
  'join_requested',
  'join_approved',
  'join_rejected',
  'room_snapshot',
  'participant_joined',
  'participant_left',
  'contribution_added',
  'contribution_withdrawn',
  'bindings_updated',
  'assistant_stream_started',
  'assistant_stream_delta',
  'assistant_stream_done',
  'assistant_stream_aborted',
  'room_error',
] as const;

export const multiplayerStateValues = ['idle', 'hosting', 'joining', 'joined'] as const;

export const multiplayerEventTypeSchema = z.enum(multiplayerEventTypeValues);
export const multiplayerStateSchema = z.enum(multiplayerStateValues);

export const roomParticipantSchema = z.object({
  participant_id: z.string().default(''),
  nickname: z.string().default(''),
  is_host: z.boolean().optional().default(false),
  connected_at: z.number().optional().default(0),
  character_name: z.string().optional().default(''),
  character_avatar: z.string().optional().default(''),
  character_card: z.unknown().optional().default({}),
}).passthrough();

export const multiplayerEnvelopeSchema = z.object({
  type: multiplayerEventTypeSchema,
  room_id: z.string().default(''),
  seq: z.number().optional().default(0),
  payload: z.unknown().optional().default({}),
  sent_at: z.number().optional().default(0),
}).passthrough();

export const roomStatusSchema = z.object({
  state: multiplayerStateSchema.default('idle'),
  room_id: z.string().nullable().optional().default(null),
  address: z.string().nullable().optional().default(null),
  port: z.number().nullable().optional().default(null),
  participant_id: z.string().nullable().optional().default(null),
  nickname: z.string().nullable().optional().default(null),
  is_host: z.boolean().optional().default(false),
}).passthrough();

export const joinRequestedSchema = z.object({
  request_id: z.string().default(''),
  nickname: z.string().default(''),
  character_name: z.string().optional().default(''),
  character_avatar: z.string().optional().default(''),
  requested_at: z.number().optional().default(0),
}).passthrough();

export const joinApprovedSchema = z.object({
  participant_id: z.string().default(''),
  nickname: z.string().default(''),
}).passthrough();

export const joinRejectedSchema = z.object({
  message: z.string().default(''),
}).passthrough();

export const contributionAddedSchema = z.object({
  contribution_id: z.string().default(''),
  room_round_id: z.string().default(''),
  participant_id: z.string().default(''),
  nickname: z.string().default(''),
  content: z.string().default(''),
  pending: z.boolean().optional().default(true),
  sent_at: z.number().optional().default(0),
}).passthrough();

export const contributionWithdrawnSchema = z.object({
  contribution_id: z.string().default(''),
  participant_id: z.string().default(''),
  withdrawn_at: z.number().optional().default(0),
}).passthrough();

export const assistantStreamSchema = z.object({
  assistant_message_key: z.string().default(''),
  delta: z.string().optional().default(''),
  reasoning: z.string().optional().default(''),
}).passthrough();

export const roomSnapshotSchema = z.object({
  room_id: z.string().default(''),
  session_key: z.string().default(''),
  session_file: z.string().default(''),
  scope_id: z.string().default(''),
  scope_name: z.string().default(''),
  bindings: sessionBindingsSchema.optional().default({}),
  participants: z.array(roomParticipantSchema).optional().default([]),
  payload: chatPayloadSchema.optional().default([]),
}).passthrough();

export const roomErrorSchema = z.object({
  message: z.string().default(''),
}).passthrough();

export const startHostDtoSchema = z.object({
  room_id: z.string().default(''),
  session_key: z.string().default(''),
  scope_id: z.string().default(''),
  scope_name: z.string().default(''),
  session_file: z.string().default(''),
  nickname: z.string().default(''),
  character_name: z.string().optional().default(''),
  character_avatar: z.string().optional().default(''),
  character_card: z.unknown().optional().default({}),
  port: z.number().optional(),
});

export const joinRoomDtoSchema = z.object({
  address: z.string().default(''),
  nickname: z.string().default(''),
  character_name: z.string().optional().default(''),
  character_avatar: z.string().optional().default(''),
  character_card: z.unknown().optional().default({}),
});

export const submitContributionDtoSchema = z.object({
  content: z.string().default(''),
});

export const withdrawContributionDtoSchema = z.object({
  contribution_id: z.string().default(''),
});

export const approveJoinDtoSchema = z.object({
  request_id: z.string().default(''),
  accept: z.boolean().default(false),
});

export const broadcastEventDtoSchema = z.object({
  type: multiplayerEventTypeSchema,
  payload: z.unknown().optional().default({}),
});

export type MultiplayerEventType = z.infer<typeof multiplayerEventTypeSchema>;
export type MultiplayerState = z.infer<typeof multiplayerStateSchema>;
export type RoomParticipant = z.infer<typeof roomParticipantSchema>;
export type MultiplayerEnvelope = z.infer<typeof multiplayerEnvelopeSchema>;
export type RoomStatus = z.infer<typeof roomStatusSchema>;
export type JoinRequestedPayload = z.infer<typeof joinRequestedSchema>;
export type JoinApprovedPayload = z.infer<typeof joinApprovedSchema>;
export type JoinRejectedPayload = z.infer<typeof joinRejectedSchema>;
export type ContributionAddedPayload = z.infer<typeof contributionAddedSchema>;
export type ContributionWithdrawnPayload = z.infer<typeof contributionWithdrawnSchema>;
export type AssistantStreamPayload = z.infer<typeof assistantStreamSchema>;
export type RoomSnapshot = z.infer<typeof roomSnapshotSchema>;
export type RoomErrorPayload = z.infer<typeof roomErrorSchema>;
export type StartHostDto = z.infer<typeof startHostDtoSchema>;
export type JoinRoomDto = z.infer<typeof joinRoomDtoSchema>;
export type SubmitContributionDto = z.infer<typeof submitContributionDtoSchema>;
export type WithdrawContributionDto = z.infer<typeof withdrawContributionDtoSchema>;
export type ApproveJoinDto = z.infer<typeof approveJoinDtoSchema>;
export type BroadcastEventDto = z.infer<typeof broadcastEventDtoSchema>;

File: src-tauri/src/domain/models/multiplayer.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MultiplayerModeState {
    Idle,
    Hosting,
    Joining,
    Joined,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct RoomDescriptor {
    pub room_id: String,
    pub session_key: String,
    pub scope_id: String,
    pub scope_name: String,
    pub session_file: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct RoomParticipant {
    pub participant_id: String,
    pub nickname: String,
    #[serde(default)]
    pub is_host: bool,
    #[serde(default)]
    pub connected_at: u64,
    #[serde(default)]
    pub character_name: String,
    #[serde(default)]
    pub character_avatar: String,
    #[serde(default)]
    pub character_card: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct JoinRequestEvent {
    pub request_id: String,
    pub nickname: String,
    #[serde(default)]
    pub character_name: String,
    #[serde(default)]
    pub character_avatar: String,
    pub requested_at: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct JoinApprovedEvent {
    pub participant_id: String,
    pub nickname: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct JoinRejectedEvent {
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ContributionEnvelope {
    pub contribution_id: String,
    pub room_round_id: String,
    pub participant_id: String,
    pub nickname: String,
    pub content: String,
    #[serde(default)]
    pub pending: bool,
    pub sent_at: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ContributionWithdrawnEnvelope {
    pub contribution_id: String,
    pub participant_id: String,
    pub withdrawn_at: u64,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct AssistantStreamEnvelope {
    pub assistant_message_key: String,
    #[serde(default)]
    pub delta: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct RoomSnapshotEnvelope {
    pub room_id: String,
    pub session_key: String,
    pub session_file: String,
    pub scope_id: String,
    pub scope_name: String,
    #[serde(default)]
    pub bindings: Value,
    #[serde(default)]
    pub participants: Vec<RoomParticipant>,
    #[serde(default)]
    pub payload: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct MultiplayerStatusDto {
    pub state: MultiplayerModeState,
    pub room_id: Option<String>,
    pub address: Option<String>,
    pub port: Option<u16>,
    pub participant_id: Option<String>,
    pub nickname: Option<String>,
    pub is_host: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct MultiplayerEnvelope {
    #[serde(rename = "type")]
    pub event_type: String,
    pub room_id: String,
    #[serde(default)]
    pub seq: u64,
    #[serde(default)]
    pub sent_at: u64,
    #[serde(default)]
    pub payload: Value,
}

impl MultiplayerEnvelope {
    pub fn new(
        event_type: impl Into<String>,
        room_id: impl Into<String>,
        seq: u64,
        sent_at: u64,
        payload: Value,
    ) -> Self {
        Self {
            event_type: event_type.into(),
            room_id: room_id.into(),
            seq,
            sent_at,
            payload,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct StartHostDto {
    pub room_id: String,
    pub session_key: String,
    pub scope_id: String,
    pub scope_name: String,
    pub session_file: String,
    pub nickname: String,
    #[serde(default)]
    pub character_name: String,
    #[serde(default)]
    pub character_avatar: String,
    #[serde(default)]
    pub character_card: Value,
    pub port: Option<u16>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct JoinRoomDto {
    pub address: String,
    pub nickname: String,
    #[serde(default)]
    pub character_name: String,
    #[serde(default)]
    pub character_avatar: String,
    #[serde(default)]
    pub character_card: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct SubmitContributionDto {
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct WithdrawContributionDto {
    pub contribution_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ApproveJoinDto {
    pub request_id: String,
    pub accept: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct BroadcastEventDto {
    #[serde(rename = "type")]
    pub event_type: String,
    #[serde(default)]
    pub payload: Value,
}
